File: Android/Sources/UserControlLibraries/BaiduMaps/src/main/java/com/genexus/android/controls/maps/baidu/MapLocation.java
package com.genexus.android.controls.maps.baidu;

import android.util.Pair;

import com.genexus.android.core.controls.maps.common.IMapLocation;
import com.baidu.mapapi.model.LatLng;

import java.util.ArrayList;
import java.util.List;

class MapLocation implements IMapLocation {
	private final LatLng mLatLng;

	public MapLocation(LatLng latLng) {
		mLatLng = latLng;
	}

	public MapLocation(double latitude, double longitude) {
		this(new LatLng(latitude, longitude));
	}

	public static MapLocation from(LatLng latLng) {
		if (latLng != null)
			return new MapLocation(latLng);
		else
			return null;
	}

	public static List<MapLocation> listFrom(List<Pair<Double, Double>> latLngs) {
		ArrayList<MapLocation> list = new ArrayList<>();
		for (Pair<Double, Double> latLng : latLngs)
			list.add(new MapLocation(new LatLng(latLng.first, latLng.second)));

		return list;
	}

	public static List<LatLng> listToLatLng(List<IMapLocation> locations) {
		ArrayList<LatLng> list = new ArrayList<>();
		for (IMapLocation location : locations)
			list.add(((MapLocation) location).getLatLng());

		return list;
	}

	static List<List<LatLng>> toLatLngLists(List<List<IMapLocation>> list) {
		List<List<LatLng>> latLngLists = new ArrayList<>();
		for (List<IMapLocation> aList : list)
			latLngLists.add(listToLatLng(aList));

		return latLngLists;
	}

	@Override
	public boolean equals(Object o) {
		if (super.equals(o))
			return true;

		if (!(o instanceof MapLocation))
			return false;

		MapLocation other = (MapLocation) o;
		return mLatLng.equals(other.mLatLng);
	}

	@Override
	public int hashCode() {
		return mLatLng.hashCode();
	}

	@Override
	public String toString() {
		return mLatLng.toString();
	}

	LatLng getLatLng() {
		return mLatLng;
	}

	@Override
	public double getLatitude() {
		return mLatLng.latitude;
	}

	@Override
	public double getLongitude() {
		return mLatLng.longitude;
	}
}

File: Android/Sources/UserControlLibraries/HMSMapKitLib/src/main/java/com/genexus/android/controls/maps/hms/MapLocation.kt
package com.genexus.android.controls.maps.hms

import android.util.Pair
import com.genexus.android.core.controls.maps.common.IMapLocation
import com.huawei.hms.maps.model.LatLng

internal class MapLocation(val latLng: LatLng) : IMapLocation {

    constructor(latitude: Double, longitude: Double) : this(LatLng(latitude, longitude)) {}

    override fun equals(other: Any?): Boolean {
        if (super.equals(other)) return true
        if (other !is MapLocation) return false
        return latLng == other.latLng
    }

    override fun hashCode(): Int {
        return latLng.hashCode()
    }

    override fun toString(): String {
        return latLng.toString()
    }

    override fun getLatitude(): Double {
        return latLng.latitude
    }

    override fun getLongitude(): Double {
        return latLng.longitude
    }

    companion object {
        fun from(latLng: LatLng?): MapLocation? {
            return if (latLng != null) MapLocation(latLng) else null
        }

        fun listFromLatLng(latLngs: List<LatLng>): List<IMapLocation> {
            val list = ArrayList<IMapLocation>()
            for (latLng in latLngs) list.add(MapLocation(latLng))
            return list
        }

        fun listFrom(latLngs: List<Pair<Double, Double>>): List<MapLocation> {
            val list = ArrayList<MapLocation>()
            for (latLng in latLngs) list.add(MapLocation(LatLng(latLng.first, latLng.second)))
            return list
        }

        fun listToLatLng(locations: List<IMapLocation>): List<LatLng> {
            val list = ArrayList<LatLng>()
            for (location in locations) list.add((location as MapLocation).latLng)
            return list
        }

        fun toLatLngList(mapLocationList: List<MapLocation?>): List<LatLng> {
            val latLngList: MutableList<LatLng> = ArrayList()
            for (mapLocation in mapLocationList) latLngList.add(mapLocation!!.latLng)
            return latLngList
        }

        fun toLatLngLists(list: List<List<IMapLocation>>): MutableList<List<LatLng>> {
            val latLngLists = mutableListOf<List<LatLng>>()
            for (aList in list)
                latLngLists.add(listToLatLng(aList))

            return latLngLists
        }
    }
}
